File: platforms/forge-1.12.2/src/main/java/heckerpowered/lethal/platform/adapter/item/ItemStackPersistentData.java
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 heckerpowered
 */

package heckerpowered.lethal.platform.adapter.item;

import heckerpowered.bridge.resources.Identifier;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.common.util.Constants;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class ItemStackPersistentData {
    private ItemStackPersistentData() {
    }

    @Nullable
    public static Long getLong(@NotNull ItemStack stack, @NotNull Identifier key) {
        final NBTTagCompound tag = stack.getTagCompound();
        final String name = key.asString();
        if (tag == null || !tag.hasKey(name, Constants.NBT.TAG_LONG)) {
            return null;
        }

        return tag.getLong(name);
    }

    public static void setLong(@NotNull ItemStack stack, @NotNull Identifier key, long value) {
        getOrCreateTag(stack).setLong(key.asString(), value);
    }

    @Nullable
    public static Double getDouble(@NotNull ItemStack stack, @NotNull Identifier key) {
        final NBTTagCompound tag = stack.getTagCompound();
        final String name = key.asString();
        if (tag == null || !tag.hasKey(name, Constants.NBT.TAG_DOUBLE)) {
            return null;
        }

        return tag.getDouble(name);
    }

    public static void setDouble(@NotNull ItemStack stack, @NotNull Identifier key, double value) {
        getOrCreateTag(stack).setDouble(key.asString(), value);
    }

    public static void remove(@NotNull ItemStack stack, @NotNull Identifier key) {
        final NBTTagCompound tag = stack.getTagCompound();
        if (tag == null) {
            return;
        }

        tag.removeTag(key.asString());
        if (tag.isEmpty()) {
            stack.setTagCompound(null);
        }
    }

    @NotNull
    private static NBTTagCompound getOrCreateTag(@NotNull ItemStack stack) {
        final NBTTagCompound existingTag = stack.getTagCompound();
        if (existingTag != null) {
            return existingTag;
        }

        final NBTTagCompound tag = new NBTTagCompound();
        stack.setTagCompound(tag);
        return tag;
    }
}

File: platforms/forge-1.12.2/src/main/kotlin/heckerpowered/lethal/platform/adapter/item/ItemStackPersistentData.kt
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 heckerpowered
 */

package heckerpowered.lethal.platform.adapter.item

import heckerpowered.bridge.resources.Identifier
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.common.util.Constants

object ItemStackPersistentData {
    @JvmStatic
    fun getLong(stack: ItemStack, key: Identifier): Long? {
        val tag = stack.tagCompound ?: return null
        val name = key.asString()
        return if (tag.hasKey(name, Constants.NBT.TAG_LONG)) tag.getLong(name) else null
    }

    @JvmStatic
    fun setLong(stack: ItemStack, key: Identifier, value: Long) {
        getOrCreateTag(stack).setLong(key.asString(), value)
    }

    @JvmStatic
    fun getDouble(stack: ItemStack, key: Identifier): Double? {
        val tag = stack.tagCompound ?: return null
        val name = key.asString()
        return if (tag.hasKey(name, Constants.NBT.TAG_DOUBLE)) tag.getDouble(name) else null
    }

    @JvmStatic
    fun setDouble(stack: ItemStack, key: Identifier, value: Double) {
        getOrCreateTag(stack).setDouble(key.asString(), value)
    }

    @JvmStatic
    fun remove(stack: ItemStack, key: Identifier) {
        val tag = stack.tagCompound ?: return
        tag.removeTag(key.asString())
        if (tag.isEmpty) stack.tagCompound = null
    }

    private fun getOrCreateTag(stack: ItemStack): NBTTagCompound {
        val existingTag = stack.tagCompound
        if (existingTag != null) return existingTag

        val tag = NBTTagCompound()
        stack.tagCompound = tag
        return tag
    }
}
